### PR TITLE
kodi: add Wayland support

### PR DESCRIPTION
**Pull request by minijackson**
###### Motivation for this change

Kodi has had support (again) for Wayland since v18.0. It would be nice to ship it in NixOS.

I mimicked Firefox for Wayland's attribute path for Kodi with plugins:

- New "plain" Kodi: `kodiPlainWayland`
- `kodiPlainWayland` gets wrapped with plugins into `kodi-wayland`

I'm open to suggestions / discussion, particularly on 2 things:

- The `libxkbcommon` build dependency is specified as `libxkbcommon.dev`, otherwise `kodiPlainWayland` won't build for me (cmake doesn't find xcbcommon), and I don't know why :-/
- Some GNU/Linux distributions ship a mix between `kodi` and `kodi-wayland`: the `/bin/kodi` script already selects the right Kodi to launch if in a Wayland or X11 environment. So if we have a derivation which compiles both `kodi` + `kodiPlainWayland`, and "copy" `${kodiPlainWayland}/lib/kodi/kodi-wayland` into `${kodi}/lib/kodi/`, we should have something that is both native to X11 and Wayland!

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/video/kodi/default.nix [hunk 1]
```diff
@@ -153,7 +158,7 @@ in stdenv.mkDerivation rec {
       which
       pkgconfig gnumake
       autoconf automake libtool # still needed for some components. Check if that is the case with 18.0
-    ];
+    ] ++ lib.optional useWayland [ wayland-protocols ];
```
Comment (jtojnar) [inline pkgs/applications/video/kodi/default.nix]: Does this really belong to `nativeBuildInputs`?

Comment (minijackson) [reply pkgs/applications/video/kodi/default.nix]: As I understand it, `wayland-protocols` are just xml files that describes the different protocols used in Wayland. They are converted to code at build time (usually with wayland-scanner for C/C++ headers).

**Comment by worldofpeace** [comment]
@minijackson Any reason why you closed this?

**Comment by minijackson** [comment]
Absolutely not, this was purely a mistake, sorry about that ^^'

**Comment by worldofpeace** [comment]
> Absolutely not, this was purely a mistake, sorry about that ^^'
> 
> Thank you for noticing and telling me!

You're welcome :smile: 

I was considering that you closed this since it's been two weeks without a review.

Perhaps @cpages would be interested in reviewing, you can also ask in a discourse post or in
#nixos on freenode.

Else fails, I should be able to in a couple of days. (this was on my sights).

File: pkgs/applications/video/kodi/default.nix [hunk 2]
```diff
@@ -13,7 +13,8 @@
 , libmpeg2, libsamplerate, libmad
 , libogg, libvorbis, flac, libxslt
 , lzo, libcdio, libmodplug, libass, libbluray
-, sqlite, mysql, nasm, gnutls, libva, libdrm, wayland
+, sqlite, mysql, nasm, gnutls, libva, libdrm
+, wayland, wayland-protocols, waylandpp, libxkbcommon
```
Comment (cpages) [inline pkgs/applications/video/kodi/default.nix]: These should be optional now that wayland is optional

File: pkgs/applications/video/kodi/default.nix [hunk 3]
```diff
@@ -144,7 +146,10 @@ in stdenv.mkDerivation rec {
     ++ lib.optional  sambaSupport    samba
     ++ lib.optional  udevSupport     udev
     ++ lib.optional  usbSupport      libusb
-    ++ lib.optional  vdpauSupport    libvdpau;
+    ++ lib.optional  vdpauSupport    libvdpau
+    ++ lib.optional  useWayland [
+      wayland waylandpp libxkbcommon.dev
```
Comment (cpages) [inline pkgs/applications/video/kodi/default.nix]: Please add a comment if we cannot sort this out

**Comment by cpages** [comment]
I just did a quick review. I could not try it, but other than my minor suggestions, looks good to me. I'd go ahead and push this to have a working kodi for wayland. We can always try to have a mixed build later on.

Btw, I'll be away for a few days, so @worldofpeace you can merge when you think it's ok. Thanks!

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build kodi-wayland

**Review by worldofpeace — DISMISSED** [review]
Can you add `wayland` to the derivation name when `useWayland` is `true`?

**Comment by minijackson** [comment]
Done!

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build kodi-wayland

##### sorry ofborg :smile:

**Review by worldofpeace — APPROVED** [review]
I've built and ran this with wayland :sparkle:

**Comment by worldofpeace** [comment]
Thank you @minijackson

**Pull request merged**